lib/Dialect/Polynomial/Transforms/NTTRewrites.td

```diff
@@ -2,43 +2,45 @@
 #define LIB_DIALECT_POLYNOMIAL_TRANSFORMS_NTTREWRITES_TD_
 
 include "lib/DRR/Utils.td"
-include "lib/Dialect/Polynomial/IR/PolynomialOps.td"
+include "mlir/Dialect/Polynomial/IR/Polynomial.td"
 include "mlir/Dialect/Arith/IR/ArithOps.td"
 include "mlir/IR/PatternBase.td"
 
 def GetRingAttr : NativeCodeCall<
-      "(dyn_cast<PolynomialType>($0.getType())).getRing()">;
+      "(dyn_cast<::mlir::polynomial::PolynomialType>($0.getType())).getRing()">;
 
 def InputTensorType : NativeCodeCall<
-      "RankedTensorType::get({$0.ideal().getDegree()},"
+      "RankedTensorType::get({$0.getPolynomialModulus().getPolynomial().getDegree()},"
       " IntegerType::get($_builder.getContext(),"
-      "  ($0.coefficientModulus() - 1).getActiveBits()),"
+      "  ($0.getCoefficientModulus().getValue() - 1).getActiveBits()),"
       " $0)">;
 
 def IntermediateTensorType : NativeCodeCall<
-      "RankedTensorType::get({$0.ideal().getDegree()},"
+      "RankedTensorType::get({$0.getPolynomialModulus().getPolynomial().getDegree()},"
       " IntegerType::get($_builder.getContext(),"
-      "  2 * ($0.coefficientModulus() - 1).getActiveBits()),"
+      "  2 * ($0.getCoefficientModulus().getValue() - 1).getActiveBits()),"
       " $0)">;
 
 def CreateCModConstant : NativeCodeCall<
       "$_builder.create<arith::ConstantOp>($0.getLoc(), $2,"
       " DenseElementsAttr::get($2,"
-      "  $1.coefficientModulus().sextOrTrunc($2.getElementTypeBitWidth())))">;
+      "  $1.getCoefficientModulus().getValue().sextOrTrunc($2.getElementTypeBitWidth())))">;
 
-// Computes (n & (n - 1)) == 0.
 def HasDegreePowerOfTwo : Constraint<
-    CPred<"APInt(64, (dyn_cast<PolynomialType>($0.getType())).getRing()"
-          ".ideal().getDegree()).isPowerOf2()">,
+    CPred<"APInt(64, (dyn_cast<::mlir::polynomial::PolynomialType>($0.getType())).getRing()"
+          ".getPolynomialModulus().getPolynomial().getDegree()).isPowerOf2()">,
     "rings are NTT compatible">;
 
+def Nullptr
+  : NativeCodeCall<"nullptr">;
+
 def NTTRewritePolyMul : Pattern<
   (Polynomial_MulOp:$mulOp $p1, $p2),
   [
     // Transform to NTT point-value representation
-    (Polynomial_NTTOp:$p1NTT $p1,
+    (Polynomial_NTTOp:$p1NTT $p1, (Nullptr),
```

Comment:
Is the `(Nullptr)` referring to a null root attribute? I actually didn't realize that the pattern match goes like (Op:$name, $input, $attr)

Reply:
It only does that if the attribute is explicitly listed on the op, which it is upstream. It should just correspond exactly to the order of `ins` in the tablegen definition of the op.